### ISSUE TITLE
fetchMore and fetchPolicy: cache-and-network does not work together

### ISSUE THREAD

**Issue by alex-golubtsov** [apollographql/apollo-client#6916]
**Intended outcome:**
In case of `fetchMore` additional request should not be executed or should use last variables (passed to `fetchMore`)

**Actual outcome:**
The request executes with wrong variables

**How to reproduce the issue:**
```tsx
function MyComponent() {
  const q = useQuery(query, { variables: { offset: 0 }, fetchPolicy: 'cache-and-network' });
  ...
  <button id="button" onClick={() => q.fetchMore({ variables: { offset: 10 } })}>Click me</button>
}
```

That's what is happening:
1. `<MyComponent />` is rendered
           => request to the server is executed with `{ offset: 0 }`
2. Button `#button` clicked
           => request to the server is executed with `{ offset: 10 }`
           => request to the server is executed with `{ offset: 0 }`

In other words the request, which fetches fresh data from the server uses variables passed to the hook, but to to `fetchMore`

**Versions**

```yml
  System:
    OS: Windows 10 10.0.19041
  Binaries:
    Node: 12.18.0
    Yarn: 1.22.4
    npm: 6.14.4
  Browsers:
    Chrome: 84.0.4147.125
  npmPackages:
    @apollo/client: ~3.1.2 => 3.1.3
    apollo: ~2.30.2 => 2.30.2
```

**Comment by DanielMarkiel** [comment]
@alex-golubtsov I believe you're describing a related issue to the one mentioned here -> https://github.com/apollographql/apollo-client/issues/6907

**Comment by khitrenovich** [comment]
See also #6327 and #6833, especially [this comment](https://github.com/apollographql/apollo-client/issues/6327#issuecomment-672041197).

Looks like the proper policy combination in 2021 for paging+cache is 
```
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first',
```

**Comment by fracmak** [comment]
Any reason why that isn’t the default setting?

**Comment by khitrenovich** [comment]
My guess would be that the system is too flexible for "one size fits all" defaults. 
You can make it default for your own setup using the recipe from [this comment](https://github.com/apollographql/apollo-client/issues/6833#issuecomment-679446789).

**Comment by Akryum** [comment]
Having the same issue in Apollo Client 3.3.13
Using `fetchPolicy: 'network-only'` and `notifyOnNetworkStatusChange: true`, calling `fetchMore` will trigger an unwanted additionnal network fetch which will mess up with the type policy `merge`.

**Comment by Akryum** [comment]
Reproduction: https://codesandbox.io/s/apolloclient-beta-fetchmore-bug-forked-8mplu?file=/src/App.js

Click multiple times on the load more button:

![image](https://user-images.githubusercontent.com/2798204/113430176-015da500-93da-11eb-99c3-773d5790f7f6.png)

Indeed apollo client does a "phantom" request after the fetch more one:

![image](https://user-images.githubusercontent.com/2798204/113430293-30741680-93da-11eb-99a3-8a1b39a7de38.png)

![image](https://user-images.githubusercontent.com/2798204/113430300-34079d80-93da-11eb-8112-70ba2710d251.png)

**Comment by Akryum** [comment]
Note: still an issue in Apollo Client 3.4.0-beta.19

**Comment by ivands** [comment]
Any update on people trying to fix this???

**Comment by Akryum** [comment]
I tried again on the codesandbox with 3.4.0-rc.23. The issue is kinda still here on the network side: the second phantom/unnecessary request with `first: 1` is still made, but it doesn't call the `merge` (so it doesn't break pagination).
So the bug halfway-fixed! :wrench:

**Comment by Akryum** [comment]
Seems it's still not working on my real project with 3.4.0-rc.23. Still having two merges with the second one being the initial query before the `fetchMore`. :sob:

**Comment by Akryum** [comment]
Digging into the source code...

The unwanted fetch seems to be caused by this line:
https://github.com/apollographql/apollo-client/blob/336337dbedc108f13035f85eacf5dde1b47ddfd8/src/core/ObservableQuery.ts#L333
Which in turn calls this with the initial variables/options:
https://github.com/apollographql/apollo-client/blob/336337dbedc108f13035f85eacf5dde1b47ddfd8/src/core/Reobserver.ts#L52

On 3.4 the lines are all in `ObservableQuery.ts` but the logic is the same:
https://github.com/apollographql/apollo-client/blob/7856c187ebf89df80c446834ab238e5475e6114d/src/core/ObservableQuery.ts#L383
And the fetch:
https://github.com/apollographql/apollo-client/blob/7856c187ebf89df80c446834ab238e5475e6114d/src/core/ObservableQuery.ts#L658

Not sure what a concast is though.

**Comment by Akryum** [comment]
If I remove the reobserve from `fetchMore`, there is another one called as a side-effect of this `writeQuery`:
https://github.com/apollographql/apollo-client/blob/7856c187ebf89df80c446834ab238e5475e6114d/src/core/ObservableQuery.ts#L372

![image](https://user-images.githubusercontent.com/2798204/127324381-4824a3b6-e2c2-4f95-9255-b7bf0cbf29a5.png)

This `writeQuery` call will lead to `setDiff` being called as well, but more importantly `notify`:
https://github.com/apollographql/apollo-client/blob/7856c187ebf89df80c446834ab238e5475e6114d/src/core/QueryInfo.ts#L205

![image](https://user-images.githubusercontent.com/2798204/127324759-2d1d2c38-f53d-4774-8222-a9567317f862.png)

This then calls a `reobserve` too, causing a new request with the initial variables/options to be sent:
https://github.com/apollographql/apollo-client/blob/7856c187ebf89df80c446834ab238e5475e6114d/src/core/QueryInfo.ts#L233

Then the result from `fetchMore` is reverted by the new `fetch`.

**Comment by Akryum** [comment]
There is probably an architectural issue if a `notify()` causes a full new `fetch`?

**Comment by brainkim** [comment]
Maybe related? https://github.com/apollographql/apollo-client/issues/8541

**Comment by rcbevans** [comment]
I've just spent ~3.5h debugging my code, then Apollo, investigating what appears to be the same issue as described, in a lazy loading list view.

Fetch more is correctly fetching and merging the next page of data, but a second request without the pagination token was then being sent and replacing the data in merge, causing the page content to reset and reset the scrollbar to the end, tiggering a refetch of the second page again, repeating infinitely

@khitrenovich suggestion to set

```ts
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first',
```

appears to suppress the second call and things now seem to work as expected.

What's frustrating is this query is only used in one place, so I'm not sure what exactly is subscribed to be notified which could be triggering this behavior.

**Comment by Akryum** [comment]
> @khitrenovich suggestion to set
> 
> ```ts
>     fetchPolicy: 'cache-and-network',
>     nextFetchPolicy: 'cache-first',
> ```
> 
> appears to suppress the second call and things now seem to work as expected.

Unfortunately this trick doesn't work with `fetchPolicy: 'network-only'` :sob:

**Comment by lrojas94** [comment]
I can also confirm this behavior, but I can't use the workaround. I really need `cache-and-network` to trigger as soon as my variables change (**every time** the variables change). It seems the `fetchPolicy + nextFetchPolicy` doesn't really do it for me :/ Is this planned at all? seems like it's been more than a year

**Comment by benjamn** [comment]
After thinking more about this, I agree with @Akryum the extra network fetch after `fetchMore` is not a reasonable consequence of using `network-only` or `cache-and-network` fetch policies.

Yes, you may be able to work around the problem with `nextFetchPolicy: "cache-first"`, but I no longer think using `nextFetchPolicy` should be necessary in this case. As long as the cache read triggered by `fetchMore` is complete (in the `diff.complete` sense), I think we should deliver that cache result immediately, and skip the network. In other words, reacting to a `fetchMore` request is not one of the situations where we want/need to reapply the original query's `fetchPolicy` literally, so I think it should be safe to ignore the `fetchPolicy` here, and just deliver the result.

I'm still investigating the consequences of these changes, but I think/hope we can remove the `this.reobserve()` from the `finally` after `fetchMore`, and change the `QueryInfo` code as follows:
```diff
commit 5c378d1d07a33ab4de0c14d2f6d68d48297ef9ec
Author: Ben Newman <ben@apollographql.com>
Date:   Tue Aug 17 16:00:56 2021 -0400

    Avoid full reobservation in QueryInfo listener when diff.complete.

diff --git a/src/core/QueryInfo.ts b/src/core/QueryInfo.ts
index ea23f374a..7f5b6ce5d 100644
--- a/src/core/QueryInfo.ts
+++ b/src/core/QueryInfo.ts
@@ -235,7 +235,8 @@ export class QueryInfo {
         // full reobservation, since oq.reobserve might make a network
         // request, and we don't want to trigger network requests for
         // optimistic updates.
-        if (this.getDiff().fromOptimisticTransaction) {
+        const diff = this.getDiff();
+        if (diff.complete || diff.fromOptimisticTransaction) {
           oq["observe"]();
         } else {
           oq.reobserve();
```
I hope to have more answers soon!

**Comment by kamilmsoprasteria** [comment]
Hi @benjamn 
any chance that this fix you posted in https://github.com/apollographql/apollo-client/issues/6916#issuecomment-901466918 will go into some near version? 3.5?

**Comment by sebastienbarre** [comment]
For those landing here after much searching and debugging, this is still an issue as of 3.4.16.
This still seems to be the "workaround" for now:
```
fetchPolicy: 'cache-and-network',
nextFetchPolicy: 'cache-first',
```

**Comment by justinhandley** [comment]
Just following this - setting the fetchPolicy as recommended does not seem to stop my app from doing a whole query when a new subscription item is received and therefore duplicating the entire cache.  I'm using the incredibly ugly hack 

`    return incoming.length > 3 ? [...incoming] : [...existing, ...incoming]`

All subscriptions return all values PLUS  the new one, which is always more than 3 (the number we get on fetchMore).  So if this is a fetchmore we merge, and otherwise we just write the incoming - since the subscription query on the front end is already merging in the new data correctly, the incoming in that case is what we need.

**Comment by videni** [comment]
what is the solution  at the moment ? still have this issue with `  @apollo/client": "^3.5.8`, I have disabled local cache.

```
 const { loading, error, data, fetchMore} = useQuery(GET_CUSTOMER_TICKETS, {
    variables: {
      first: DEFAULT_PAGE_SIZE,
      last: DEFAULT_PAGE_SIZE
    },
    fetchPolicy: 'no-cache',
    nextFetchPolicy: 'no-cache',
  });
```

**Comment by alamothe** [comment]
@videni The whole world is pretty much waiting on this and other similar issues. Hopefully it will make it to 3.6 release. 

The last good version of Apollo Client was 2.6.10 and we are still on this version.

**Comment by jairmedeiros** [comment]
Any update? In this moment I can't use fetch more or create an infinite scroll 😭.

There are some workaround to make this fetchMore?

**Comment by jonathan-guerrero-milli** [comment]
@jairmedeiros 
Based on [this comment](https://github.com/apollographql/apollo-client/issues/7496#issuecomment-939290629) it seems like a workaround is to `useState` for the variables and modify them instead of calling`fetchMore` or `refetch`.

`useQuery` should requery based on changed variables, see the linked comment for how they did it with `useLazyQuery`.

It probably requires a refactor though to handle the loading/error states (our code used to wait for/catch the `fetchMore` promise to show a loader/message in the footer of our SectionList)

**Comment by benjamn** [comment]
Some good news, at long last: using @Akryum's [reproduction](https://codesandbox.io/s/apolloclient-beta-fetchmore-bug-forked-8mplu?file=/src/App.js) and `@apollo/client@3.6.0-beta.10`, I see no extra network requests after the `fetchMore` calls (thanks to #9504). These changes have just landed (today) on the `release-3.6` branch, so we are actively looking for help testing them.

While this change seems like the right behavior for most use cases, there's a chance it will be disruptive for code that was previously relying on `fetchMore` triggering extra network requests (hard to imagine, but certainly possible). Please help us test/validate these changes by updating your projects with `npm i @apollo/client@beta` when/if you have the chance! Your help will get v3.6 released sooner (in April).

**Comment by nateql** [comment]
Was closed on the team board so I am closing it here.

**Comment by alamothe** [comment]
Ho about github.com/apollographql/apollo-client/issues/7485 and https://github.com/apollographql/apollo-client/issues/6760, will those be addressed as well?

**Comment by hatched-kade** [comment]
I had a similar issue to @zakton5. I originally though it was a `fetchPolicy` issue, but what was actually causing the behavior was a custom merge function for a field that originally looked like 
```
merge: (existing, incoming) => incoming || existing
```
That field could be null for some of my data while I was doing a `fetchMore`, so `existing=undefined` and `incoming=null`. My merge function would return `undefined`, which must've triggered some silent error causing apollo to retry the original query, with the network status going from 3 -> 1. 

Changing my merge to be:
```
merge: (existing, incoming) => {
  if (!incoming && existing) {
    return existing;
  }
  return incoming;
}
```
caused the issue to stop happening.

**Comment by engrwaqasali** [comment]
> @jairmedeiros Based on [this comment](https://github.com/apollographql/apollo-client/issues/7496#issuecomment-939290629) it seems like a workaround is to `useState` for the variables and modify them instead of calling`fetchMore` or `refetch`.
> 
> `useQuery` should requery based on changed variables, see the linked comment for how they did it with `useLazyQuery`.
> 
> It probably requires a refactor though to handle the loading/error states (our code used to wait for/catch the `fetchMore` promise to show a loader/message in the footer of our SectionList)

if you useState then you will lose the apollo cache updating your data whenever the data is changed

**Comment by okanji** [comment]
I had the same problem as @hatched-kade. This is my strategy where I incorporated the check that was provided above:

```js
const cache = new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          projectsPaginated: {
            keyArgs: false,
            merge(existing, incoming, { args }) {
              const merged = existing ? existing.projects.slice(0) : [];
              const { offset = 0 } = args;

              if (!incoming && existing) {
                return existing;
              }
              
              if (incoming) {
                if (args) {
                  for (let i = 0; i < incoming.projects.length; ++i) {
                    merged[offset + i] = incoming.projects[i];
                  }
                } else {
                  throw Error("args not defined");
                }
              }

              return { ...incoming, projects: merged };
            },
          },
        },
      },
    },
  });
```

This is my caching strategy for my query `projectsPaginated`. This is the type defenition:

```js
    privateCloudProjectsPaginated(
      offset: Int!
      limit: Int!
      search: String
      sort: String
      sortOrder: SortOrder
    ): ProjectPagination! 
```
